pysteps/blending/steps.py

```diff
@@ -65,1723 +64,2857 @@
-    Prediction System (STEPS) method.
-
-    Parameters
-    ----------
-    precip: array-like
-      Array of shape (ar_order+1,m,n) containing the input precipitation fields
-      ordered by timestamp from oldest to newest. The time steps between the
-      inputs are assumed to be regular.
-    precip_models: array-like
-      Either raw (NWP) model forecast data or decomposed (NWP) model forecast data.
-      If you supply decomposed data, it needs to be an array of shape
-      (n_models,timesteps+1) containing, per timestep (t=0 to lead time here) and
-      per (NWP) model or model ensemble member, a dictionary with a list of cascades
-      obtained by calling a method implemented in :py:mod:`pysteps.cascade.decomposition`.
-      If you supply the original (NWP) model forecast data, it needs to be an array of shape
-      (n_models,timestep+1,m,n) containing precipitation (or other) fields, which will
-      then be decomposed in this function.
-
-      Depending on your use case it can be advantageous to decompose the model
-      forecasts outside beforehand, as this slightly reduces calculation times.
-      This is possible with :py:func:`pysteps.blending.utils.decompose_NWP`,
-      :py:func:`pysteps.blending.utils.compute_store_nwp_motion`, and
-      :py:func:`pysteps.blending.utils.load_NWP`. However, if you have a lot of (NWP) model
-      members (e.g. 1 model member per nowcast member), this can lead to excessive memory
-      usage.
-
-      To further reduce memory usage, both this array and the ``velocity_models`` array
-      can be given as float32. They will then be converted to float64 before computations
-      to minimize loss in precision.
+# TODO: compare old and new version of the code, run a benchmark to compare the two
+# TODO: look at the documentation and try to improve it, lots of things are now combined together
 
-      In case of one (deterministic) model as input, add an extra dimension to make sure
-      precip_models is four dimensional prior to calling this function.
-    velocity: array-like
-      Array of shape (2,m,n) containing the x- and y-components of the advection
-      field. The velocities are assumed to represent one time step between the
-      inputs. All values are required to be finite.
-    velocity_models: array-like
-      Array of shape (n_models,timestep,2,m,n) containing the x- and y-components
-      of the advection field for the (NWP) model field per forecast lead time.
-      All values are required to be finite.
 
-      To reduce memory usage, this array
-      can be given as float32. They will then be converted to float64 before computations
-      to minimize loss in precision.
-    timesteps: int or list of floats
-      Number of time steps to forecast or a list of time steps for which the
-      forecasts are computed (relative to the input time step). The elements of
-      the list are required to be in ascending order.
+@dataclass
```

Comment:
I just realized, based on your comment that the config variables should be immutable, that this should be a frozen dataclass to make that explicit. This will simply disallow you from mutating the config object after creation.
```suggestion
@dataclass(frozen=True)
```

Reply:
The problem with that fix is that during __check_inputs some variables are set to be a dict or something else if nothing is provided by the user. There are two suggestions:
- Do not used the frozen label
- Define the standard of the variables to be the assigned values in __check_inputs. This might not work for all scenarios as the assignments might be conditional.

Reply:
I think option 2 should work, I see that there is one case where the config default value is conditional. I think all the other cases could just be default values. For this one case I also think that we should be able to work around this pretty easilly, but this does raise the question to me if this derived value should be a part of the config class?
Is `n_models` not always equal to `self.__precip_models.shape[0]` and thus it should not be part of a config that you can also provide yourself?
We could for example move this value to the state in stead of the config?

```
            self.__config.climatology_kwargs = dict(
                {"n_models": self.__precip_models.shape[0]}
            )
```

Reply:
There is actually one more place where the config changes during the nowcasting with is this line. I also think we should be able to do without this line right? Because `xy_coordinates` is already stored in the params so it does not need to be copied to the config.
```
self.__config.extrapolation_kwargs["xy_coords"] = self.__params.xy_coordinates
```

Reply:
One more thing I would like to add is that while `frozen=True` makes the config object immutable any dict that is in it is still mutable, so the line from my previous comment would still be allowed. Ideally also that line would be disallowed. Apparently you can use `MappingProxyType` to make the dictionaries immutable: https://adamj.eu/tech/2022/01/05/how-to-make-immutable-dict-in-python/

Reply:
Ok well immutable dicts opens its own can of worms so maybe that is too much for now, but `frozen=True` is definitely attainable. I tried this out myself: https://github.com/pySTEPS/pysteps/pull/447/files and the tests still seem to succeed. I did remove the `{"n_models": self.__precip_models.shape[0]}` part of the code, but as far as I can tell it is used nowhere in this file, do you know if this is used anywhere?

Reply:
the "n_models" is used but in one of the helper functions of the sklill_scores: clim_regr_values(). So just removing it is not an option. However, if I make the standard a dict and a dict is still mutable, I don't think we have an issue, right?

Reply:
I just checked this and my tests seem to fail: The reason is, the original forecast function defines a lot of these variables that are assigned to be a dict to be None. Thus, my "field(default_factory=dict)" assigmnet as the default value does not apply as something is provided by the user. I could however change all the mutated variables over to the params class and then we would get rid of that issue... 
What do you think?

Reply:
I think the movement to the __params is not a bad idea with running multiple forecasts in mind. Lets say the __config would be mutable and the first forecast ran has x members while a later forecast has y members. That would currently not be dealt with correctly by the code. Is we move the dict assignments to the __params, that issue would go away.

Reply:
I just saw your commit where you made the config object immutable. Looks good!